### PR TITLE
Introduce a new aggregation module

### PR DESCRIPTION
**Pull request by martijnvg**
Also this PR moves the `adjacency_matrix` to this new module. 
This serves as an example how to move the other aggregations from the server module into this new module for aggregations.

The new module name is `aggregations`.
The new module will use the `org.elasticsearch.aggregations.bucket` package for all bucket aggregations. (`org.elasticsearch.search.aggregations.bucket` is already taken by server module).
We can change the naming as part of reviewing this PR.

Relates to #90283

### CODE REVIEW

File: modules/aggregations/build.gradle [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+apply plugin: 'elasticsearch.internal-yaml-rest-test'
+apply plugin: 'elasticsearch.yaml-rest-compat-test'
```
Comment (martijnvg) [inline modules/aggregations/build.gradle:9]: this should pull yaml tests defined in module into bwc yaml tests.

**Comment by martijnvg** [comment]
Opening this PR as draft in order for this change to pass PR CI. Probably some issues will occur, that I didn't foresee locally.

File: server/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java [hunk 2]
```diff
@@ -138,7 +137,6 @@ public class AggregationsTests extends ESTestCase {
         new InternalDateRangeTests(),
```
Comment (martijnvg) [inline server/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java:137]: I think we want this test also in the new module? As we move over aggregations we remove entries here and add it to the equivalent test case in  aggregations module?

File: modules/aggregations/src/main/java/module-info.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+module org.elasticsearch.aggs {
+    requires org.elasticsearch.base;
+    requires org.elasticsearch.server;
+    requires org.elasticsearch.xcontent;
+    requires org.apache.logging.log4j;
```
Comment (nik9000) [inline modules/aggregations/src/main/java/module-info.java:13]: We might should just use ES's logging directly while we're moving.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationInspectionHelper.java [hunk 4]
```diff
@@ -82,7 +82,7 @@ public static boolean hasValue(UnmappedSampler agg) {
         return false;
     }
 
-    public static boolean hasValue(InternalAdjacencyMatrix agg) {
+    public static boolean hasValue(MultiBucketsAggregation agg) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationInspectionHelper.java:85]: I've never really understood the point of this method. It seems like we could assert this in the tests if we wanted.

File: server/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java (hunk 2 above)
Comment (nik9000) [reply server/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java:137]: How do we think this should look in the end? Would there be an `AggregationsTests` in `test:framework` that has no aggs listed? That feels maybe not great. But neither does having `test:framework` depend on the module. Maybe it'd make sense to move this class to a `aggregations:test` module that other aggregation extensions depend on. That'd pick up all these tests.

Comment (nik9000) [reply server/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java:137]: But in the short term removing this from the list is fine.

Comment (martijnvg) [reply server/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java:137]: > Maybe it'd make sense to move this class to a aggregations:test module that other aggregation extensions depend on. That'd pick up all these tests.

Yes, I think when enough tests (50%?) have moved we should move this test to the aggregation module.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationInspectionHelper.java (hunk 4 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationInspectionHelper.java:85]: Right, this is used in tests (`GeoGridAggregatorTestCase`) and sql (`MetricAggExtractor`). I think this should be moved to sql (looks like the biggest user of this class?) and the one usage in `GeoGridAggregatorTestCase` can be replaced with adding a helper method that does `agg.getBuckets().stream().anyMatch(bucket -> bucket.getDocCount() > 0)`.

Let's do this in a followup change?

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationInspectionHelper.java:85]: :+1:

File: server/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java (hunk 2 above)
Comment (nik9000) [reply server/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java:137]: :+1:

**Review by nik9000 — APPROVED** [review]

File: rest-api-spec/build.gradle [hunk 5]
```diff
@@ -35,6 +35,7 @@ artifacts {
 
 testClusters.configureEach {
   module ':modules:mapper-extras'
+  module ':modules:aggregations'
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply rest-api-spec/build.gradle:41]: I added comment to remind us: 71a3530154276105ff304e41159be0f9104b50de

**Pull request merged**